docker/daemon.go

```diff
@@ -150,16 +154,31 @@ func mainDaemon() {
 	job.Setenv("TlsCert", *flCert)
 	job.Setenv("TlsKey", *flKey)
 	job.SetenvBool("BufferRequests", true)
-	err := job.Run()
+
+	// The serve API job never exits unless an error occurs
+	// We need to start it as a goroutine and wait on it so
+	// daemon doesn't exit
+	serveAPIWait := make(chan struct{})
+	go func() {
+		defer close(serveAPIWait)
+
+		if err := job.Run(); err != nil {
+			log.Errorf("Error from ServeAPI: %v", err)
+		}
+	}()
 
 	// Wait for the daemon startup goroutine to finish
 	// This makes sure we can actually cleanly shutdown the daemon
-	log.Infof("waiting for daemon to initialize")
-	<-daemonWait
-	eng.Shutdown()
+	log.Debug("waiting for daemon to initialize")
+	err := <-daemonInitWait
 	if err != nil {
+		eng.Shutdown()
```

Comment:
actually we don't want to at this point, as we know we have an error during daemon init and we need to shutdown cleanly. The serve API handler will happily sit there waiting to be told to "acceptconnections", even though the daemon failed initialization, which means we would wait forever without stopping.